solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol

```diff
@@ -0,0 +1,1362 @@
+        view
+        validReserve(_reserveToken)
+        returns (uint32)
+    {
+        return PPM_RESOLUTION / 2;
+    }
+
+    /**
+      * @dev returns the balance of a given reserve token
+      *
+      * @param _reserveToken    reserve token contract address
+      *
+      * @return the balance of the given reserve token
+    */
+    function reserveBalance(IERC20Token _reserveToken)
+        public
+        view
+        returns (uint256)
+    {
+        uint256 reserveId = __reserveIds[_reserveToken];
+        require(reserveId != 0, "ERR_INVALID_RESERVE");
+        return readReserveBalance(reserveId);        
+    }
+
+    /**
+      * @dev returns the balances of both reserve tokens
+      *
+      * @return the balances of both reserve tokens
+    */
+    function reserveBalances()
+        public
+        view
+        returns (uint256, uint256)
+    {
+        return readReserveBalances(1, 2);        
+    }
+
+    /**
+      * @dev converts a specific amount of source tokens to target tokens
+      * can only be called by the bancor network contract
+      *
+      * @param _sourceToken source ERC20 token
+      * @param _targetToken target ERC20 token
+      * @param _amount      amount of tokens to convert (in units of the source token)
+      * @param _trader      address of the caller who executed the conversion
+      * @param _beneficiary wallet to receive the conversion result
+      *
+      * @return amount of tokens received (in units of the target token)
+    */
+    function convert(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount, address _trader, address payable _beneficiary)
+        public
+        override
+        payable
+        protected
+        only(BANCOR_NETWORK)
+        returns (uint256)
+    {
+        // validate input
+        require(_sourceToken != _targetToken, "ERR_SAME_SOURCE_TARGET");
+
+        return doConvert(_sourceToken, _targetToken, _amount, _trader, _beneficiary);
+    }
+
+    /**
+      * @dev returns the conversion fee for a given target amount
+      *
+      * @param _targetAmount  target amount
+      *
+      * @return conversion fee
+    */
+    function calculateFee(uint256 _targetAmount) internal view returns (uint256) {
+        return _targetAmount.mul(conversionFee) / PPM_RESOLUTION;
+    }
+
+    /**
+      * @dev reads the stored reserve balance for a given reserve id
+      *
+      * @param _reserveId   reserve id
+    */
+    function readReserveBalance(uint256 _reserveId) internal view returns (uint256) {
+        return (__reserveBalances >> ((_reserveId - 1) * 128)) & MAX_UINT128;        
+    }
+
+    /**
+      * @dev reads the stored reserve balances
+      *
+      * @param _sourceId    source reserve id
+      * @param _targetId    target reserve id
+    */
+    function readReserveBalances(uint256 _sourceId, uint256 _targetId) internal view returns (uint256, uint256) {
+        require(_sourceId + _targetId == 3, "ERR_INVALID_RESERVES");
```

Comment:
Possible values of these IDs are 0 (for non existing reserves), and then 1 and 2.
So 1+2 and 2+1 are the only valid scenarios, and the alternative would cost more gas, in common utility function used in several critical paths.

Reply:
either the function should do validation and then it should do proper validation, or it assumes that the caller does the validation and then it shouldn't do any - the current implementation doesn't follow any of these options.

Reply:
what I mean is that the caller can pass MAX_UINT and 4 and the check will pass with invalid ids.

Reply:
No external caller can pass `reserveId`.
It is passed only between functions in this contract, and the only possible values that it can attain are 0, 1 and 2.

Reply:
by "external caller" I mean a different function in the same contract (of course) that calls this function. and it can pass these values: MAX_UINT, 4 and the check will pass.

Reply:
Changed `_sourceId + _targetId == 3` to `(_sourceId == 1 && _targetId == 2) || (_sourceId == 2 && _targetId == 1)`.
Note that it adds more code and increases the gas-cost (no upside IMO).